src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Input/Commands/Ordering/OrderingHelper.cs

```diff
@@ -112,142 +75,104 @@ public static bool TryMoveDown(Project project, IProjectTree projectTree)
-            // Empty folders do not have a valid display order.
-            // In the future, if we want to order empty folders, meaning they have a valid display order, 
-            //     this will need to be extended to handle that behavior.
+            var includes = GetEvaluatedIncludes(projectTree);
 
-            // This recursively tries to find a node to work with. If there is none at all, this will break on project root.
-            var currentTarget = target;
-            while(nodeToAddAbove == null && currentTarget.IsFolder && !GetChildren(currentTarget).Any())
-            {
-                currentTarget = currentTarget.Parent;
-                nodeToAddAbove = GetFirstChild(currentTarget);
-            }
+            var elements = new List<ProjectItemElement>();
 
-            if (nodeToAddAbove != null && nodes.Any())
+            foreach (var include in includes)
             {
-                var didAllMoveSuccessfully = true;
+                // GetItemsByEvaluatedInclude is efficient and uses a MultiDictionary underneath.
+                //     It uses this: new MultiDictionary<string, ProjectItem>(StringComparer.OrdinalIgnoreCase);
+                var item = project.GetItemsByEvaluatedInclude(include).FirstOrDefault();
 
-                foreach (var node in nodes)
+                // We only care about adding one item associated with the evaluated include.
+                if (item?.Xml is ProjectItemElement element)
                 {
-                    if (!TryMoveAbove(project, node, nodeToAddAbove))
-                    {
-                        didAllMoveSuccessfully = false;
-                    }
+                    elements.Add(element);
                 }
-
-                return didAllMoveSuccessfully;
             }
 
-            return false;
+            return elements.ToImmutableArray();
         }
 
         /// <summary>
-        /// This moves the nodes above the target's first child. 
-        /// If there are no children with a valid display order of the target or the target is an empty folder, it will try the parent target's first child.
+        /// If the project gets any additional items from the callback, this will return those added items.
         /// </summary>
-        public static Task<bool> TryMoveNodesToTopAsync(ConfiguredProject configuredProject, IEnumerable<IProjectTree> nodes, IProjectTree target)
+        public static async Task<ImmutableArray<ProjectItemElement>> AddItems(ConfiguredProject configuredProject, IProjectAccessor accessor, Func<Task> addItems)
         {
-            return ModifyProjectAsync(configuredProject, project => TryMoveNodesToTop(project, nodes, target));
+            var previousIncludes = await accessor.OpenProjectForReadAsync(configuredProject, project =>
+            {
+                return project.AllEvaluatedItems.Select(x => x.EvaluatedInclude).ToImmutableHashSet(StringComparer.OrdinalIgnoreCase);
+            }).ConfigureAwait(true);
```

Comment:
Why ConfigureAwait(true)?

The project could change between these two OpenProjectForReadAsync drastically, what happens if it does?